Make series a separate model

hughrun opened the pull request:
## Description

This PR moves book series to a pair of new models: `Series` and `SeriesBook`. This will allow more consistency around series, and for books to be matched to series regardless of whether they share an author. Basically it makes series "things, not strings".

A `Series` has basic `BookData` to better enable us to de-duplicate them, although only series imported from Inventaire are likely to have anything beyond a name. A `SeriesBook` connects a Series to a `Book`, including a `series_number`: it's the instantiation of a Book being part of a series.

When saving a book, users will be prompted to confirm whether it is part of an existing series, in the same way we check for existing authors.

All importers will now convert series information to a `SeriesBook` and `Series` where possible. If there is a matching series but no other book in the series with a matching author, the series information is left on the incoming book and users will be prompted to confirm the series information manually on the book view page.

Series can be edited, including updating the `seriesnumber` on one or more `seriesbooks` - users need to have permission to edit books to do this.

Creating a data migration proved to be challenging, so I have instead created an admin command to convert "legacy" series to Series and SeriesBooks where possible. This will need to be run by admins manually.

I have tried to make this backwards-compatible but it's entirely possibly that I've missed something.

I also added an atomic block to the user export model which is entirely unrelated except that I made a mistake when I was testing, and without this block I got a bunch of obscure errors.

This was a bit of a beast to put together so there may well be parts that could be more efficient or that I've missed - happy for any feedback!

- Related Issue #3681 #2959
- Closes #3579 
- Closes #3306 
- Closes #2991 

## What type of Pull Request is this?
<!-- Check all that apply -->

- [ ] Bug Fix
- [x] Enhancement
- [ ] Plumbing / Internals / Dependencies
- [ ] Refactor

## Does this PR change settings or dependencies, or break something?
<!-- Check all that apply -->

- [ ] This PR changes or adds default settings, configuration, or .env values
- [ ] This PR changes or adds dependencies
- [ ] This PR introduces other breaking changes

### Details of breaking or configuration changes (if any of above checked)


## Documentation
<!--
Documentation for users, admins, and developers is an important way to keep the BookWyrm community welcoming and make Bookwyrm easy to use.
Our documentation is maintained in a separate repository at https://github.com/bookwyrm-social/documentation
-->

<!-- Check all that apply -->

- [x] New or amended documentation will be required if this PR is merged
- [ ] I have created a matching pull request in the Documentation repository
- [ ] I intend to create a matching pull request in the Documentation repository after this PR is merged

<!-- Amazing! Thanks for filling that out. Your PR will need to have passing tests and happy linters before we can merge
You will need to check your code with `black`, `pylint`, and `mypy`, or `./bw-dev formatters`
-->

### Tests
<!-- Check one -->

- [ ] My changes do not need new tests
- [x] All tests I have added are passing
- [ ] I have written tests but need help to make them pass
- [ ] I have not written tests and need help to write them

hughrun committed:
Create Series and SeriesBook models

hughrun committed:
add connector logic and fix templates and views

hughrun committed:
finalise connectors

update templates
add migration

hughrun committed:
refactor series models and templates

also adds tests

hughrun committed:
bunch of changes and tests

hughrun committed:
formatting

hughrun committed:
final commit for series

hughrun committed:
add migration

hughrun committed:
use SearchVector in connectors

- match to existing series using a search vector instead of exact matching
- add series listing to author pages
- clean up a few other details

hughrun committed:
add management command and improve connector series matching

hughrun committed:
fix upgrade_series admin command

hughrun committed:
fix linters hopefully

ilkka-ollakka commented on bookwyrm/views/books/edit_book.py:
```diff
@@ -220,6 +235,40 @@ def add_authors(request, data):
+    matches = (
+        models.Series.objects.annotate(search=vector)
+        .annotate(rank=SearchRank(vector, series, normalization=32))
+        .filter(rank__gt=0.019)
```
I think this should be 0.19 instead of 0.019 ?

mouse-reeve committed:
Merge branch 'main' into series-model

hughrun committed:
fix rank match value for series matches

ilkka-ollakka commented:
> hmm. Not sure why this `NotifyInviteRequest` test is failing, I didn't do anything to notifications. I suspect something about ordering/concurrency in the test suite?

I didn't manage to reproduce the failure locally either.

ilkka-ollakka reviewed (approved):
I think this looks good. The failing test is something that could be tried to reproduce but shouldn't block merging in my opinion.

hughrun committed:
Resolve conflicts and merge 'main' into series-model

Resolve conflicts.

hughrun committed:
fix series title search in ConfirmEditBook

Q object query was wrong.
Also fixed an except block I messed up when merging.

mouse-reeve committed:
Merge branch 'main' into series-model

mouse-reeve commented on bookwyrm/templates/book/series.html:
```diff
@@ -2,35 +2,79 @@
+			<div class="is-flex-grow-1 mb-3">
+				<meta itemprop="isPartOf" content="{{series.name}}">
+				<meta itemprop="name" content="{{book.title}}">
+				<span class="subtitle">{% if seriesbook.series_number %}{% blocktrans with series_number=seriesbook.series_number %}Book #{{ series_number }}{% endblocktrans %}{% endif %}</span>
```
Since series number doesn't have to be a number, should the pound sign be here? It looks odd if the value of `series_number` is a word. I'm trying bolding the series number field and removing the #

mouse-reeve commented on bookwyrm/templates/book/series.html:
```diff
@@ -2,35 +2,79 @@
+			<div class="column is-one-quarter">
+				{% if series.inventaire_id or series.wikidata or series.isfdb %}
+				<div class="box">
+					<h4 class="title is-4">{% trans "Identifiers" %}</h4>
```
I changed this to "External links" to be consistent with the similar box on the author page

mouse-reeve reviewed (changes requested):
I submitted fixes for all these comments on your repo, hopefully that helps??? But the two errors I found unfortunately remain

mouse-reeve committed:
Hughrun series model (#10)

* Adds migration files needed for merge

* Sorts series by number

Also fixes a blocktrans error

* Small wording and formatting tweaks to seris view

dato commented on bookwyrm/management/commands/upgrade_series.py:
```diff
@@ -0,0 +1,77 @@
+
+    help = "Turn legacy series data into Series and SeriesBook objects"
+
+    # pylint: disable=no-self-use,unused-argument
```
```suggestion
```

dato reviewed (commented):
I took a look at this PR, I have some minor comments and questions (all suggestions are optional).

Thanks!!

hughrun commented on bookwyrm/models/book.py:
```diff
@@ -891,3 +927,67 @@ def preview_image(instance, *args, **kwargs):
+    book = fields.ForeignKey(
+        "Book", on_delete=models.CASCADE, related_name="seriesbooks"
+    )
+    series_number = fields.CharField(max_length=255, blank=True, null=True)
```
Yes, it should be CharField, thanks for the pickup.

hughrun commented on bookwyrm/models/book.py:
```diff
@@ -891,3 +927,67 @@ def preview_image(instance, *args, **kwargs):
+    @property
+    def collection_queryset(self):
+        """list of books for this series, overrides OrderedCollectionMixin"""
+        return SeriesBook.objects.filter(series=self).order_by("-series_number")
```
Yes! But it actually wasn't working anyway because `OrderedCollectionMixin` was missing, so thanks for pointing this out.

hughrun committed:
clean up

- fix some tests
- remove user/actor from Series
- fix bugs with saving series manually

hughrun committed:
Merge branch 'main' into series-model

hughrun committed:
Merge branch 'main' into series-model

hughrun commented on bookwyrm/tests/activitypub/test_series.py:
```diff
@@ -0,0 +1,262 @@
+
+    def test_serialize_series(self):
+        """check presence of series fields"""
+        activity = self.series.to_activity()
```
Series doesn't need an actor, and also can't serialise back to books and seriesbooks, at least for now. So I've removed `actor`/user from the Series model, and also removed this test.

hughrun committed:
basically undo previous changes to models

- turns out we need 'user' on Series as well
- fix reverse values for activitypub

hughrun committed:
fix series model

- fix upgrade_series command and add test
- make series a proper ordered collection
- fix various test problems
- prevent duplicate seriesbooks
- fix problems in edit_book template due to series management
- fix issues with series pages

hughrun committed:
Merge branch 'main' into series-model

hughrun committed:
Merge branch 'main' into series-model

ilkka-ollakka commented on bookwyrm/management/commands/upgrade_series.py:
```diff
@@ -0,0 +1,86 @@
+                series=series,
+                book=book.parent_work,
+                series_number=book.series_number,
+            )
```
and here also.

mouse-reeve reviewed (changes requested):
I'm sorry to be the bearer of this news, but now when I import a book _without_ a series from Inventaire, it creates a blank string series that has a url linked from the book page, but the link leads to a 404

hughrun commented:
hmm, that seems suboptimal! I must have broken the logic somewhere.

hughrun committed:
Merge branch 'main' into series-model

hughrun committed:
fix inventaire series and clean up

- fixes issue where imported books from Inventaire created series with no name
- uses django's get_or_create() for simpler code
- rationalises migrations
- adds tests

hughrun committed:
Merge branch 'main' into series-model

hughrun committed:
Merge branch 'main' into series-model

hughrun commented:
@mouse-reeve @ilkka-ollakka this is ready to look at again.

ilkka-ollakka commented:
I tested out this as dev instance and importing somewhat large goodreads export file in.

For me it seems to give exception when I manually try to move one book from 'Wayfarer, #3' series to 'Wayfarer' series, so misparsing on some data. Adding book into existing series works, but removing the 'Wayfarer, #3' series from the book seems to give followin:

```
Exception Value: SeriesBook matching query does not exist.
Raised during: 	bookwyrm.views.books.edit_book.EditBook
```
<img width="3272" height="1286" alt="image" src="https://github.com/user-attachments/assets/ef17dc7a-bb2e-43c3-ad4a-599a870cc7e4" />


I think it happens because I already selected to remove the existing series when added new one, but it somehow stayed with the book and next removal try gives error?

hughrun commented:
@ilkka-ollakka I'll add some error handling here but can you give me any more information about your workflow to produce this error? I can't replicate it, unless I have two edit pages for the same book open in different windows/tabs.

ilkka-ollakka commented on bookwyrm/templates/book/edit/edit_book_form.html:
```diff
@@ -147,6 +124,57 @@ <h2 class="title is-4">
+                                <label class="label mb-2" for="remove_series">
+                                    <span class="tag is-link is-light is-medium">
+                                        <a href="{{ sb.series.local_path }}">{{ sb.series.name }}</a>
+                                        <input class="ml-4" type="checkbox" name="remove_series" id="id_remove_series_{{sb.series.id}}" value="{{ sb.series.id }}" {% if sb.series.id|stringformat:"i" in remove_series %}checked{% endif %} aria-describedby="desc_remove_series_{{sb.series.id}}">
```
should the value be `sb.id` instead of `sb.series.id` in here? In the code it seems to seek SeriesBook.id=this-value in remove_series list.

hughrun replied:
omg.

You are absolutely correct. I changed the way the models worked a couple of times and I must have missed this when I was cleaning up. I guess when I was testing, the seriesbook and the series had the same ID so it seemed to be working. Thanks for your attention to detail, @ilkka-ollakka, this would have been pretty disastrous to release!

hughrun committed:
use seriesbook id rather than series id for remove_series

The form template incorrectly passed series id instead of seriesbook id.
Now it doesn't.

ilkka-ollakka commented:
Adding and removing book from series now works for me, I seemed to hit some issue after I ran 'deduplicate_book_data', after running that upgrade_series gives me integrityerror:
```
django.db.utils.IntegrityError: null value in column "book_id" of relation "bookwyrm_seriesbook" violates not-null constraint
DETAIL:  Failing row contains (1356, 2026-04-25 10:53:03.270295+00, 2026-04-25 10:53:03.270303+00, null, , null, 1332, 2)
```

I'll give it another run later to see if I can reproduce the issue and pin it down more precisely.

hughrun commented:
@ilkka-ollakka Is it possible you have editions with no `parent_work`? I think that might be what causes this. I will add a check for this, but it would be good to have confirmation whether this is the cause.

hughrun committed:
fix orphans choking on upgrade_series command

ilkka-ollakka commented:
I'm unable to reproduce the issue with recent commit.

One thing that I noticed, after running 'upgrade_series' command, books that don't have series, get found edition amount of empty series linked to.

<img width="1130" height="1374" alt="kuva" src="https://github.com/user-attachments/assets/982a7840-6f68-4e7e-bb47-a179ce8e3c16" />

In edit book view those are all listed as empty series as

<img width="1283" height="1401" alt="kuva" src="https://github.com/user-attachments/assets/664819b9-975d-4e25-b773-6b31050b9b9b" />


Empty series amounts in all the books that I checked seems to match edition count of the book.

hughrun commented:
@ilkka-ollakka can you show me a relevant sample of the goodreads file? I'm wondering if this might be happening if series is a blank string (`""`) rather than null.

hughrun committed:
exclude series with empty strings in upgrade_series command

ilkka-ollakka commented on bookwyrm/management/commands/upgrade_series.py:
```diff
@@ -14,7 +14,7 @@ def upgrade_series_data():
     seriesbook_count = SeriesBook.objects.count()
 
-    for book in Edition.objects.exclude(series=None):
+    for book in Edition.objects.exclude(parent_work=None).exclude(series=None):
```
After I locally added also `.exclude(series="")` it stopped mapping to empty series. Seems that most of the books have blank series in db after import.

ilkka-ollakka commented:
> @ilkka-ollakka can you show me a relevant sample of the goodreads file? I'm wondering if this might be happening if series is a blank string (`""`) rather than null.

I can provide the full goodreads export if you like? But seems to be that series is blank instead of None, as you suspected.

hughrun commented:
Thanks @ilkka-ollakka! If you can't see any more issues then I think that is solved so I don't need the file.

@mouse-reeve I know you've already looked at this but do you mind doing a final check? It's a pretty big change so it would be good to have another set of eyes on it, especially given my silly mistakes @ilkka-ollakka identified.

mouse-reeve commented:
I don't know if this is a blocker, but I found this behavior a little unexpected from a user perspective:
When a book has a series , the edit page shows the option to remove the series, or add a new series. It doesn't show the option to change the existing series number. I can add a series number and it's saved but not used (because it's not associated with the existing series entry). 

The only way to change the series number, as far as I can tell, is to remove the series, and then re-add it with the number?

hughrun commented:
hmm. The expected way to change an existing series number is to go to the series page and do it there.

I agree that is not very intuitive. The problem is that a book can be in more than one series simultaneously, and in that case we don't know which series number you are changing.

This _should_ be a blocker because we already have too many obscure things like this. I'll consider some alternatives.

mouse-reeve commented:
I see! Would the simplest fix be to have a link to "edit series order" or something in the edit book page? Ideally you could edit the number on the existing series but I don't know how much that complicates things.

hughrun commented:
i'll investigate both options today. keen to get this merged, but it needs to work properly.

hughrun commented:
@mouse-reeve is this too busy?

<img width="668" height="689" alt="Screenshot 2026-05-16 at 8 42 37 AM" src="https://github.com/user-attachments/assets/80797f12-849e-4015-ad2a-df4e4007219e" />

mouse-reeve commented:
That works! Maybe if the individual series were in a `box` element and the series number on the same line as the label element?

hughrun committed:
improves series editing on edit book page

- fixes error where remove author was unchecked if confirm_mode active
- adds ability to amend series_number when editing book
- improves UI for editing series in book edit screen
- add unique_together constraint to series + series_number
- adds error message when user tries to duplicate series number in either book or series editing screen
- minor UI improvements when editing series

hughrun commented:
Ok @mouse-reeve that was a little more complicated than I expected, but it did help me to notice that there was an unrelated bug with removing authors, and I hadn't put a unique-together constraint on series+series_number.

When you have time, I think this is ready for a hopefully-last check. I've tested every combination I can think of, but you might find something I missed.

mouse-reeve committed:
Adds missing trans tags in edit_series.html

mouse-reeve committed:
Adds missing trans tag to book.html

mouse-reeve reviewed (approved):
Looks good! How exciting!! I added a couple missing translation tags but that's it.

joric3 commented:
I just want to say thanks for the work you all put in. You're awesome!
